### PR TITLE
Implement mmap syscall

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR implements the Mmap syscall. In order to do this, it
* adds support for optionally writing to registers,
* moves the heap pointer into the registers,
* adds a heap pointer increase function,
* implements the mmap syscall.

### CODE REVIEW

File: optimism/src/mips/interpreter.rs [hunk 1]
```diff
@@ -248,19 +286,19 @@ pub trait InterpreterEnv {
 
             instruction_counter + Self::constant(1)
         };
-        unsafe { self.push_register_access(idx, new_accessed.clone()) };
+        unsafe { self.push_register_access_if(idx, new_accessed.clone(), if_is_true) };
         self.add_lookup(Lookup {
             numerator: Signed {
                 sign: Sign::Pos,
-                magnitude: Self::constant(1),
+                magnitude: if_is_true.clone(),
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:293]: Will this "add" a lookup with magnitude zero when `if_is_true` is false?

Comment (mrmr1993) [reply optimism/src/mips/interpreter.rs:293]: That's the intention, yes. It's basically a 'non-contributing lookup term'.

File: optimism/src/mips/interpreter.rs [hunk 2]
```diff
@@ -805,7 +881,37 @@ pub fn interpret_rtype<Env: InterpreterEnv>(env: &mut Env, instr: RTypeInstructi
             return;
         }
         RTypeInstruction::JumpAndLinkRegister => (),
-        RTypeInstruction::SyscallMmap => (),
+        RTypeInstruction::SyscallMmap => {
+            let requested_alloc_size = env.read_register(&Env::constant(5));
+            let size_in_pages = {
```
Comment (dannywillems) [inline optimism/src/mips/interpreter.rs:886]: I don't understand this. Why `instruction`?

Comment (mrmr1993) [reply optimism/src/mips/interpreter.rs:886]: Oopsie! Fixed

File: optimism/src/mips/interpreter.rs [hunk 3]
```diff
@@ -805,7 +881,37 @@ pub fn interpret_rtype<Env: InterpreterEnv>(env: &mut Env, instr: RTypeInstructi
...
+                // FIXME: Requires a range check
+                let pos = env.alloc_scratch();
+                unsafe { env.bitmask(&requested_alloc_size, 32, PAGE_ADDRESS_SIZE, pos) }
```
Comment (dannywillems) [inline optimism/src/mips/interpreter.rs:889]:
```suggestion
                unsafe { env.bitmask(&requested_alloc_size, PAGE_ADDRESS_SIZE, 0, pos) }
```
No?

Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:889]: (checking twice)

File: optimism/src/mips/interpreter.rs [hunk 4]
```diff
@@ -805,7 +881,37 @@ pub fn interpret_rtype<Env: InterpreterEnv>(env: &mut Env, instr: RTypeInstructi
...
+            let actual_alloc_size =
+                (size_in_pages + requires_extra_page) * Env::constant(1 << PAGE_ADDRESS_SIZE);
+            let address = env.read_register(&Env::constant(4));
+            let address_is_zero = env.is_zero(&address);
+            let old_heap_ptr = env.increase_heap_pointer(&actual_alloc_size, &address_is_zero);
+            let return_position = {
+                let pos = env.alloc_scratch();
+                env.copy(
+                    &(address_is_zero.clone() * old_heap_ptr
+                        + (Env::constant(1) - address_is_zero) * address),
```
Comment (dannywillems) [inline optimism/src/mips/interpreter.rs:905]: Should it not be the new size of the heap?

Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:905]: Ignore.

**Review by dannywillems — APPROVED** [review]

**Pull request merged**